remoting/src/main/java/org/apache/rocketmq/remoting/util/ServiceProvider.java

```diff
@@ -131,6 +132,53 @@ protected static InputStream getResourceAsStream(ClassLoader loader, String name
         return services;
     }
 
+    public static Map<String, String> loadPath(String path) {
+        LOG.info("Load path looking for a resource file of name [{}] ...", path);
+        Map<String, String> pathMap = new HashMap<String, String>();
+        try {
+            final InputStream is = getResourceAsStream(getContextClassLoader(), path);
+            if (is != null) {
+                BufferedReader reader;
+                try {
+                    reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
+                } catch (java.io.UnsupportedEncodingException e) {
+                    reader = new BufferedReader(new InputStreamReader(is));
+                }
+                String serviceName = reader.readLine();
+                while (serviceName != null && !"".equals(serviceName)) {
+                    String[] service = serviceName.split("=");
+                    if (service.length == 2) {
+                        if (pathMap.containsKey(service[0])) {
+                            continue;
+                        } else {
+                            pathMap.put(service[0], service[1]);
+                        }
+                    } else {
+                        continue;
+                    }
+                    serviceName = reader.readLine();
+                }
+                reader.close();
+            }
+        } catch (Exception ex) {
+            LOG.error("Error occurred when looking for resource file " + path, ex);
+        }
+        return pathMap;
+    }
+
+    public static <T> Map<String, T> load(String path, Class<?> clazz) {
+        LOG.info("Load map is looking for a resource file of name [{}] ...", path);
+        Map<String, T> services = new HashMap<String, T>();
+        Map<String, String> pathMaps = loadPath(path);
+        for (Map.Entry<String, String> entry : pathMaps.entrySet()) {
+            T instance = (T) createInstance(entry.getValue(), clazz);
+            if (instance != null && !services.containsKey(entry.getKey())) {
+                services.put(entry.getKey(), instance);
+            }
+        }
+        return services;
+    }
+
     public static <T> T loadClass(String name, Class<?> clazz) {
         final InputStream is = getResourceAsStream(getContextClassLoader(), name);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *is*. The resource is referenced in statements at the following lines: 184, 188, 190, 192. The resource closure statement is at line: 193. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by BufferedReader.readLine. Either a) close *is* in a try-finally block or b) close the resource by declaring *is* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).